test/metadata-extractor/src/main/java/org/elasticsearch/extractor/features/HistoricalFeaturesMetadataExtractor.java

```diff
@@ -62,23 +65,36 @@ public void generateMetadataFile(Path outputFile) {
             XContentGenerator generator = JsonXContent.jsonXContent.createGenerator(os)
         ) {
             generator.writeStartObject();
-            for (Map.Entry<NodeFeature, Version> entry : extractHistoricalFeatureMetadata().entrySet()) {
-                generator.writeStringField(entry.getKey().id(), entry.getValue().toString());
-            }
+            extractHistoricalFeatureMetadata((historical, names) -> {
+                generator.writeFieldName("historical_features");
+                generator.writeStartObject();
+                for (Map.Entry<NodeFeature, Version> entry : historical.entrySet()) {
+                    generator.writeStringField(entry.getKey().id(), entry.getValue().toString());
+                }
+                generator.writeEndObject();
+                generator.writeFieldName("feature_names");
+                generator.writeStartArray();
+                for (var entry : names) {
+                    generator.writeString(entry);
+                }
+                generator.writeEndArray();
+            });
             generator.writeEndObject();
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
     }
 
-    public Map<NodeFeature, Version> extractHistoricalFeatureMetadata() {
+    void extractHistoricalFeatureMetadata(CheckedBiConsumer<Map<NodeFeature, Version>, Set<String>, IOException> metadataConsumer)
```

Comment:
Why a biconsumer here? Why not return `Tuple<Map<,>, Set<>>`?

Reply:
I don't like tuples, they are too "opaque" - they make it difficult to tell what is what and to see types